src/universe/population.cpp

```diff
@@ -15,56 +15,75 @@
 
 #include "oldstructs.hpp"
 
+cCreature cPopulation::dummyDude;
+
 void cPopulation::import_legacy(legacy::creature_list_type old){
 	dudes.resize(60);
-	for(int i = 0; i < 60; i++)
-		dudes[i].import_legacy(old.dudes[i]);
+	for(int i = 0; i < 60; i++) {
+		dudes[i]=std::make_shared<cCreature>();
```

Comment:
Hmm, I'm not too keen on storing every cCreature as a shared pointer. What are the problems that can occur when they're killed without this change?

Reply:
In many cases, the game does not check correctly if a creature is killed. So after killing a creature (and deleting its memory), it still changes some of its members. This is why many games/saves have become corrupted :-~

**Note** In practice, using a share_ptr seems to be the easiest way to avoid this problem because when you take care of a creature, many external things can happen (which can kill it) before you are done taking care of it.

Reply:
So basically you're saying there out-of-bounds writes on the `dudes` array. Okay.

Reply:
Yes.

But there is also the case (less crash-generating) where some creatures are summoned after the death of a creature, i.e. if we continue to modify the members of the creature that has just died, we will modify some characteristics of a new summoned creature.

Reply:
Hmm… leaving stale monsters in the array indefinitely is certainly a possibility, though that creates a chance that the `dudes` array grows without bound…

Reply:
To be compatible with older compilers, the simplest solution might be to use boost::shared_ptr ( which is the ancestor of std::shared_ptr ). 

Because using std::vector<cCreature> dudes and passing a reference to dudes[2] to a method is really dangerous: if the dudes vector grows (because a monster is summoned), the reference points anywhere. 

At the very least, replacing std::vector<cCreature> dudes with a std::array<cCreature, XXX> is maybe a bit healthy (even if you have to choose XXX and add protections in the code).

Reply:
> To be compatible with older compilers, the simplest solution might be to use boost::shared_ptr ( which is the ancestor of std::shared_ptr ).

I'm choosing not to be compatible with any compiler that doesn't support (at least most of) C++11. Therefore, no need for `boost::shared_ptr`. I'm already using `std::shared_ptr` in other places.

My objection here isn't to `std::shared_ptr` but to the idea of using a shared pointer (or even a unique pointer) at all.

> Because using _std::vector< cCreature >_ dudes and passing a reference to dudes[2] to a method is really dangerous: if the dudes vector grows (because a monster is summoned), the reference points anywhere.

Ah, do things keep references (or pointers) to elements of the creature array? If so, maybe we should play it safe and use `std::deque` instead… but that said, as long as the references don't outlive a function, it should still be fine with `std::vector`…

> At the very least, replacing std::vector dudes with a std::array<cCreature, XXX> is maybe a bit healthy (even if you have to choose XXX and add protections in the code).

This is an option as a last resort, but it puts us back in the old state where creatures have a hard upper limit. I could still choose a _much higher_ upper limit than the old game had, but…

Reply:
For instance, if you look at the code:
```
void do_monster_turn() {
	cCreature *cur_monst;
....	
	for(short i = 0; i < num_monst; i++) {  // Give monsters ap's, check activity
		
		cur_monst = &univ.town.monst[i];
....
						acted_yet = monst_cast_mage(cur_monst,target);
....

```
this is problematic, because cur_monst keeps a pointer to univ.town.monst[i] and this pointer does not changed when univ.town.monst is resized (e.g., when a new monster is summoned) :-~

Reply:
Okay. I think this can be solved by using `std::deque` instead of `std::vector`, as long as it's always resized only by adding a new one on the end.

Reply:
I've changed it to a `std::deque`. Hopefully that'll be enough to fix any issues you saw, as references to elements of a deque remain valid when a new element is added to the end.